vector-stores/spring-ai-cassandra/src/main/java/org/springframework/ai/vectorstore/CassandraFilterExpressionConverter.java

```diff
@@ -0,0 +1,188 @@
+import static org.springframework.ai.vectorstore.filter.Filter.ExpressionType.LTE;
+import static org.springframework.ai.vectorstore.filter.Filter.ExpressionType.NE;
+import static org.springframework.ai.vectorstore.filter.Filter.ExpressionType.NIN;
+import static org.springframework.ai.vectorstore.filter.Filter.ExpressionType.NOT;
+import static org.springframework.ai.vectorstore.filter.Filter.ExpressionType.OR;
+import org.springframework.ai.vectorstore.filter.Filter.Group;
+import org.springframework.ai.vectorstore.filter.Filter.Key;
+import org.springframework.ai.vectorstore.filter.Filter.Value;
+import org.springframework.ai.vectorstore.filter.converter.AbstractFilterExpressionConverter;
+
+/**
+ * Converts {@link Expression} into CQL where clauses.
+ *
+ * @author Mick Semb Wever
+ */
+final class CassandraFilterExpressionConverter extends AbstractFilterExpressionConverter {
+
+	private final Map<String, ColumnMetadata> columnsByName;
+
+	public CassandraFilterExpressionConverter(Collection<ColumnMetadata> columns) {
+
+		columnsByName = columns.stream()
+			.collect(Collectors.toMap((c) -> c.getName().asInternal(), Function.identity()));
+	}
+
+	@Override
+	public String convertExpression(Filter.Expression expression) {
+		// TODO
+		// scan and collect all keys in the expression
+		// and validate we have a valid where clause
+		// rules:
+		// - if one partition column is specified, or partition columns must be
+		// - if a clustering column is specified all previous clustering columns must be
+		// specified with EQ operand
+		return super.convertExpression(expression);
+	}
+
+	@Override
+	protected void doKey(Key key, StringBuilder context) {
+		String keyName = key.key();
+		Optional<ColumnMetadata> column = getColumn(keyName);
+		Preconditions.checkArgument(column.isPresent(), "No metafield %s has been configured", keyName);
+		context.append(column.get().getName().asCql(false));
+	}
+
+	@Override
+	protected void doExpression(Filter.Expression expression, StringBuilder context) {
+		switch (expression.type()) {
+			case AND:
+				doBinaryOperation(" and ", expression, context);
+				break;
+			case OR:
+				doBinaryOperation(" or ", expression, context);
+				break;
+			case NIN:
+			case NOT:
+				throw new UnsupportedOperationException(
+						format("Expression type %s not yet implemented. Patches welcome.", expression.type()));
+			default:
+				doField(expression, context);
+				break;
+		}
+	}
+
+	private static void doOperand(ExpressionType type, StringBuilder context) {
+		switch (type) {
+			case EQ:
+				context.append(" = ");
+				break;
+			case NE:
+				context.append(" != ");
+				break;
+			case GT:
+				context.append(" > ");
+				break;
+			case GTE:
+				context.append(" >= ");
+				break;
+			case IN:
+				context.append(" IN ");
+				break;
+			// TODO SAI supports collections
+			// reach out to mck@apache.org if you'd like these implemented
```

Comment:
`CONTAINS` is a reverse of `IN`

"Is the value of my_column IN these_values" (where these_values is a collection)

versus

"Is this_value IN my_column" (where my_column is a collection)

So… i haven't thought if this would be possible.  I generally see the assumption that the left entry in an expression is a Key (and the right the Value).